bundles/auth/org.eclipse.smarthome.auth.oauth2client/src/main/java/org/eclipse/smarthome/auth/oauth2client/internal/OAuthClientServiceImpl.java

```diff
@@ -0,0 +1,356 @@
+            @NonNull String password, @Nullable String scope) throws OAuthException, IOException, OAuthError {
+        checkNotNull(username, "username must not be null");
+        checkNotNull(password, "password must not be null");
+
+        OAuthConnector connector = new OAuthConnector(httpClientFactory);
+        AccessTokenResponse accessTokenResponse = connector.grantTypePassword(persistedParams.tokenUrl, username,
+                password, persistedParams.clientId, persistedParams.clientSecret, scope,
+                persistedParams.supportsBasicAuth);
+
+        if (accessTokenResponse != null) {
+            // store it
+            storeHandler.saveAccessTokenResponse(handle, accessTokenResponse);
+        }
+        return accessTokenResponse;
+    }
+
+    @SuppressWarnings("null")
+    @Override
+    public AccessTokenResponse getAccessTokenByClientCredentials(@Nullable String scope)
+            throws OAuthException, IOException, OAuthError {
+
+        OAuthConnector connector = new OAuthConnector(httpClientFactory);
+        AccessTokenResponse accessTokenResponse = connector.grantTypeClientCredentials(persistedParams.tokenUrl,
+                persistedParams.clientId, persistedParams.clientSecret, scope,
+                Boolean.TRUE.equals(persistedParams.supportsBasicAuth));
+
+        if (accessTokenResponse != null) {
+            // store it
+            storeHandler.saveAccessTokenResponse(handle, accessTokenResponse);
+        }
+        return accessTokenResponse;
+    }
+
+    @SuppressWarnings("null")
+    @Override
+    public AccessTokenResponse refreshToken() throws OAuthException, IOException, OAuthError {
+        AccessTokenResponse lastAccessToken = storeHandler.loadAccessTokenResponse(handle);
+        if (lastAccessToken == null) {
+            throw new OAuthException(
+                    "Cannot refresh token because last access token is not available from handle: " + handle);
+        }
+        if (lastAccessToken.getRefreshToken() == null) {
+            throw new OAuthException("Cannot refresh token because last access token did not have a refresh token");
+        }
+
+        OAuthConnector connector = new OAuthConnector(httpClientFactory);
+        AccessTokenResponse accessTokenResponse = connector.grantTypeRefreshToken(persistedParams.tokenUrl,
+                lastAccessToken.getRefreshToken(), persistedParams.clientId, persistedParams.clientSecret,
+                persistedParams.scope, Boolean.TRUE.equals(persistedParams.supportsBasicAuth));
+        if (accessTokenResponse != null) {
+            // store it
+            storeHandler.saveAccessTokenResponse(handle, accessTokenResponse);
+        }
+        return accessTokenResponse;
+
+    }
+
+    @Override
+    public AccessTokenResponse getAccessTokenResponse() throws OAuthException, IOException, OAuthError {
+        checkNotNull(handle, "handle must not be null");
```

Comment:
Such invariant is really good for parameter validation. Since this method doesn't have any param, we could just add a null if check.

Reply:
Fixed - implemented in the getInstance part